pkg/object/mqttproxy/broker.go

```diff
@@ -0,0 +1,336 @@
+	b.Unlock()
+
+	client.session.updateEGName(b.egName, b.name)
+	topics, qoss, _ := client.session.allSubscribes()
+	if len(topics) > 0 {
+		err = b.topicMgr.subscribe(topics, qoss, client.info.cid)
+		if err != nil {
+			logger.Errorf("client <%v> use previous session topics <%v> to subscribe failed, err:%v", client.info.cid, topics, err)
+		}
+	}
+	client.readLoop()
+}
+
+func (b *Broker) setSession(client *Client, connect *packets.ConnectPacket) {
+	// when clean session is false, previous session exist and previous session not clean session,
+	// then we use previous session, otherwise use new session
+	prevSess := b.sessMgr.get(connect.ClientIdentifier)
+	if !connect.CleanSession && (prevSess != nil) && !prevSess.cleanSession() {
+		// prevSess.update(connect)
+		client.session = prevSess
+	} else {
+		if prevSess != nil {
+			prevSess.close()
+		}
+		client.session = b.sessMgr.newSessionFromConn(connect)
+	}
+}
+
+func (b *Broker) requestTransfer(egName, name string, data HTTPJsonData) {
+	urls, err := b.memberURL(egName, name)
+	if err != nil {
+		logger.Errorf("mqtt.requestTransfer: not find url for eg:%s, name:%s, err:%v", egName, name, err)
+		return
+	}
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		logger.Errorf("mqtt.requestTransfer: json data marshal failed, err: %v", err)
+		return
+	}
+	for _, url := range urls {
+		req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+		if err != nil {
+			logger.Errorf("mqtt.requestTransfer: make new request failed, err:%v", err)
+			return
+		}
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			logger.Errorf("mqtt.requestTransfer: http client send msg failed, err:%v", err)
+			return
+		}
+		defer resp.Body.Close()
+	}
+}
+
+func (b *Broker) sendMsgToClient(topic string, payload []byte, qos byte) {
+	subscribers, _ := b.topicMgr.findSubscribers(topic)
+	if subscribers == nil {
+		logger.Errorf("mqtt.sendMsgToClient: not find subscribers for topic %s", topic)
+		return
+	}
+
+	for clientID, subQos := range subscribers {
+		if subQos < qos {
+			return
+		}
+		sess := b.sessMgr.get(clientID)
+		if sess == nil {
+			logger.Errorf("mqtt.sendMsgToClient: session for client %s is nil", clientID)
+		} else {
+			if sess.info.EGName == b.egName && sess.info.Name == b.name {
+				sess.publish(topic, payload, qos)
+			}
+		}
+	}
+}
+
+func (b *Broker) getClient(clientID string) *Client {
+	b.RLock()
+	defer b.RUnlock()
+	if val, ok := b.clients[clientID]; ok {
+		return val
+	}
+	return nil
+}
+
+func (b *Broker) topicsPublishHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		api.HandleAPIError(w, r, http.StatusBadRequest, fmt.Errorf("suppose POST request but got %s", r.Method))
+		return
+	}
+	var data HTTPJsonData
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		api.HandleAPIError(w, r, http.StatusBadRequest, fmt.Errorf("invalid json data from request body"))
+		return
+	}
+	if data.Qos < 0 || data.Qos > 2 {
```

Comment:
`0` & `2` should be replaced with constant defined in other files.

Reply:
OK, I will fix that. Thanks a lot